File: 2690-HouseRobberIv/2690-HouseRobberIv.java
// Last updated: 6/17/2025, 10:24:49 PM
class Solution {
    
    private boolean canSteal(int[] nums, int k, int capability) {
        int steal = 0;
        int i = 0;
        int n = nums.length;

        while (i < n && steal < k) {
            if (nums[i] <= capability) {
                steal++;
                i += 2;
            } else {
                i++;
            }
        }

        return steal >= k;
    }

    public int minCapability(int[] nums, int k) {
        int n = nums.length;
        int minH = Integer.MAX_VALUE;
        int maxH = Integer.MIN_VALUE;

        for (int val : nums) {
            minH = Math.min(minH, val);
            maxH = Math.max(maxH, val);
        }

        while (minH < maxH) {
            int mid = (minH + maxH) / 2;

            if (canSteal(nums, k, mid)) {
                maxH = mid;
            } else {
                minH = mid + 1;
            }
        }

        return minH;
    }
}

File: 2690-HouseRobberIv/2690-HouseRobberIv.py
# Last updated: 6/17/2025, 10:24:54 PM
class Solution:
    def minCapability(self, nums: List[int], k: int) -> int:
        def canRob(mid):
            count, i = 0, 0
            while i < len(nums):
                if nums[i] <= mid:
                    count += 1
                    i += 2
                else:
                    i += 1
            return count >= k

        left, right = min(nums), max(nums)
        
        while left < right:
            mid = (left + right) // 2
            if canRob(mid):
                right = mid
            else:
                left = mid + 1
        
        return left